dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,154 @@
 
 namespace dwave::optimization {
 
+std::vector<ssize_t> make_concatenate_shape(std::vector<ArrayNode*> array_ptrs, ssize_t axis) {
+    // One or more arrays must be given
+    if (array_ptrs.size() < 1) {
+        throw std::invalid_argument("need at least one array to concatenate");
+    }
+
+    for (auto it = std::next(array_ptrs.begin()), stop = array_ptrs.end(); it != stop; it++) {
+
+        // Arrays must have the same number of dimensions
+        if ((*std::prev(it))->ndim() != (*it)->ndim()) {
+            throw std::invalid_argument(
+                    "all the input arrays must have the same number of dimensions," +
+                    std::string(" but the array at index ") +
+                    std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                    " has " + std::to_string((*std::prev(it))->ndim()) +
+                    " dimension(s) and the array at index " +
+                    std::to_string(std::distance(array_ptrs.begin(), it)) +
+                    " has " +
+                    std::to_string((*it)->ndim()) +
+                    " dimension(s)");
+        }
+
+        // Array shapes must be the same except for on the concatenation axis
+        for (ssize_t i = 0; i < (*it)->ndim(); i++) {
+            if (i != axis) {
+                if ( (*std::prev(it))->shape()[i] != (*it)->shape()[i] ) {
+                    throw std::invalid_argument(
+                            "all the input array dimensions for the concatenation" +
+                            std::string(" axis must match exactly, but along dimension ") +
+                            std::to_string(i) + ", the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), std::prev(it))) +
+                            " has size " +
+                            std::to_string((*std::prev(it))->shape()[i]) +
+                            " and the array at index " +
+                            std::to_string(std::distance(array_ptrs.begin(), it)) +
+                            " has size " +
+                            std::to_string((*it)->shape()[i]));
+                }
+            }
+        }
+    }
+
+    // Axis must be in range 0..ndim-1
+    // We can do this check on the first input array since
+    // we at this point know they all have the same shape
+    if (!(0 <= axis && axis < array_ptrs.at(0)->ndim())) {
```

Comment:
Performance doesn't really matter in this code, but we've already checked that `array_ptrs.size() > 0` so we should just use `[0]`. Or, if we want to be even more explicit, we can use `array_ptrs.front()->ndim()`

Reply:
fixed